### DISCUSSION TITLE
Guppyscreen todo list

### DISCUSSION THREAD

**Discussion by consp** [xblax/flashforge_ad5m_klipper_mod#134] [Development]
List of thing to do to get it properly working, suggestion are welcome.

- [x] create buildroot package
- [x] create mainsail prompt items ([backportable](https://github.com/ballaswag/guppyscreen/pull/97))
- [x] fix blanking (disable backlight)
- [x] enable changing of brightness
- [x] make evdev select the correct event (backportable)
- [x] test all commands (level, shaker etc) shaker might crash klipper
- [x] fix invisible wifi password input field ([backportable](https://github.com/ballaswag/guppyscreen/pull/70))
- [x] enable deletion of wifi items (backportable)
- [x] make sure wpa autoconnects
- [x] include or change code to use/not use the custom macos (e.g. input shaper, not a macro issue but related to system load)
- [x] disable spoolman
- [x] disable update guppy
- [x] provide proper config
- [ ] Add 'restart' / 'restart_firmware' prompt (backportable)
- [ ] Add 'save_config' prompt (backportable)

changes so far: https://github.com/consp/guppyscreen/tree/flashforge_ad5m

(backportable) -> items can be added to main branch without too much issues if the maintainer whiches. Other items are platform specific.

**Comment by ballaswag** [comment]
The custom input shaper is almost a one to one copy of that’s in klipper. It adds a structure json return for easier consumption inside Guppy Screen. It also enables displaying of the png in Guppy Screen.

**Reply by consp** [reply]
Will probably write a pass-though macro, shouldn't be too much of a problem, that way the main codebase can remain the same.

Thanks for the heads up, will save me quite some time looking into the changes :)

**Comment by consp** [comment]
Added BR2_PACKAGE_GUPPYSCREEN to the buildroot as well, needs some makefile patches since every makefile thinks it's in control (and isn't). See https://github.com/xblax/flashforge_adm5_buildroot_chroot/compare/master...feature/buildsystem

**Reply by xblax** [reply]
Ok, great! I will move that package into the main repo together with the other custom buildroot packages.
I have already started to work on that, but need to clean this up a bit.

**Reply by consp** [reply]
Sounds good, I'm not going to mess with it too much right now, getting the makefiles to work was quite an adventure enough 😅 

I did push some changes sincve it needs the libevdev dependency otherwise the build will fail due to missing headers.

**Reply by consp** [reply]
It also needs a change to the post build script as it now deploys the guppyscreen in printer_software and that is `rm -rf`'d in that file (pushed to feature branch).

**Reply by consp** [reply]
If the wifi changes get accepted I'm probably going to make it a patch and pull from the original repo, that way it is easier to stay up to date and simply patch everything.

**Comment by consp** [comment]
on wpa auto connecting ... looks like it does, but it takes quite a while. At least it looks to be that way. If anyone has more ideas about this I'd like to know.

**Reply by ballaswag** [reply]
Maybe there a way to force some client to acquire dhcp lease quicker? Maybe dhclient?

**Reply by consp** [reply]
Could be, or a race condition where the dhcp client already sends a request before the link is ready. Wpa sometimes also "waits" a while  for some reason before attempting to connect. Haven't looked much further though.

**Comment by xblax** [comment]
@consp 
I have integrated your Guppyscreen branch while working on the buildsystem (see https://github.com/xblax/flashforge_ad5m_klipper_mod/issues/32). It's still on the buildsystem branch, but will then push it to master soon.

Build worked basically on first attempt for me and memory usage really looks much better. Great work!

Should we plan to already include Guppyscreen build in the next release? (even if not, we can alreadyl integrate the changes on master).

**Reply by consp** [reply]
Probably yes, I already updated a bit which should be present since it pulls the branch not the commit. Few minor changes to do but those are not critical. I do have some issues with wifi only sometimes: it gets an ip, connects but does not respond to pings and not to any tcp/ip traffic. It does respond to arp requests though.

**Comment by consp** [comment]
Checklist for panel tests and known issues:

## Test checklist

- [x] homing (including moving)
- [x] load/unload (note: extrude/retract wait for temperature so be patient)
- [x] fans
- [x] led
- [x] print
- [ ] printer settings
  - [x] mesh -> note: uses `BED_MESH_CALIBRATE`
  - [x] finetune
  - [ ] input shaper -> crashes first time same as klipperscreen due to eboard timeout, second time it works. Looks like numpy hogs too much memory on first load (maybe load it ahead of time?)
  - [x] belts 
  - [x] limits
  - [x] tmc autotune (disabled)
  - [x] tmc metrics (does nothing as it's not enabled)
- [x] settings
  - [x] wifi
  - [x] system settings
  - [x] restart guppy
  - [x] restart klipper
  - [x] restart firmware

## Known issues
- some menu's might work a bit different from klipperscreen, keep this in mind
- Input shaper will stop the first time due to excessive memory consumption triggering an eboard timeout
- FIRMWARE_RESET message might not show up after input shaper error, you must use fluidd or mainsail in that case.
- RESET message might not show up after input shaper error, you must use fluidd or mainsail in that case.
- "Printer initializing" message might be stuck on screen for a while. (infinity is also a while, if you have this, please let it be known)
- Pressing the power button will not wake up the screen
- ~~Wifi will only work when ethernet is connected (yes, it's broken)~~ looks like it works now

**Reply by ballaswag** [reply]
For input shaping and belts, did you setup the guppy macros? They are wrappers around the existing shaper calibration/klippain but with some input and output changes.

In terms of load for input shaping. The k1 128mb SoC also crashes sometimes. The Creality firmware actually shipped their own c executable that does psd cal instead of relying on the python stack.

**Reply by consp** [reply]
yes, now testing everything with the guppy macros, looks like it's working once I fix the paths properly (bugs out now but it triggers everything as I expect)

For the input shaper, I'm pretty sure numpy is the root cause of this, running it without any gui leaves about 1-5MB of memory and running it with any gui it consumes it too fast for the swap to help out. Guppy is lean on memory but not that lean that it can run in the few mb's remaining.

Since python does a lot of loading during module start (first load of numpy) I'm almost positive the garbage collector fixes the memory hog after a while (too late for the first input shaper run). It "crashes" after the x test and at the start of calculations.

**Reply by ballaswag** [reply]
I also start playing with the bootlin toolchain. Was able to build a single static executable with their musl variant. From my end that's at least simpler for distributing for multiplatform.

Might not be as useful for you since you have the build procedures ported to buildroot.

**Reply by consp** [reply]
Also, don't look too much at the known issues, since it's mainly a list of things to add to the release notes since it differs a bit from how klipperscreen functions. I'd like to let users know what to expect instead :).

**Reply by consp** [reply]
> static executable with their musl variant

That will definately save a bit, we also build armhf since we have full control of the buildspace so that might differ as well. Might be another incompatability.

**Reply by ballaswag** [reply]
Tmc metrics work if you include the last 3 klippy modules here https://github.com/ballaswag/guppyscreen/tree/main/k1/k1_mods and enable them in printer.cfg.

Tmc autotune, I feel like it's somewhere a hit and miss. I didn't enable and distribute for the k1 because the tmc autotune module needed some small refactor to work on the old k1 klipper.

If this is a thing you want for your printer and it's running newer klipper, you can just install this klippy module and it should enable the button.

https://github.com/andrewmcgr/klipper_tmc_autotune

**Reply by consp** [reply]
> running newer klipper

Unfortunately that's a bit of a bind, since flashforge decided in their wisdom that GPL does not include the makefiles which have the custom configs for the STM32 clone chip they used, so for now it's v0.11 or bust (which is what is shipped).

**Reply by consp** [reply]
> Tmc autotune, I feel like it's somewhere a hit and miss. I didn't enable and distribute for the k1 because the tmc autotune module needed some small refactor to work on the old k1 klipper.

I think I'll disable them for now as it's a 'advanced user' thing and not really needed for normal operation.

**Reply by consp** [reply]
> Tmc metrics work if you include the last 3 klippy modules here https://github.com/ballaswag/guppyscreen/tree/main/k1/k1_mods and enable them in printer.cfg.

Good point! I'll have a look one the new version of the mod is released.

**Reply by consp** [reply]
@ballaswag fyi, preloading numpy at start of klipper allows the input shaper to work first time around with guppyscreen (at least in my fresh-from-boot-inputshaper-test)

edit: well ... one out of three times

**Comment by b-fitzpatrick** [comment]
I've been trying out the Guppyscreen version. Most everything is working great. @consp, looking at your fork of Guppyscreen, it appears that you added support for response actions (but not button groups, yet). I am trying to use a prompt as just a notification with only an "OK" button. That seems to work if I only interact with Guppyscreen. The prompt shows up on screen in Fluidd, and it dismisses both places if OK is tapped on screen. If I dismiss via Fluidd, it hangs Guppyscreen. First, the button text changes to gibberish, and when I tap it, it hangs.

Is there a better way to show a notification? My immediate usage is for filament runout.

**Reply by consp** [reply]
This is experimental since it's not (yet) in guppyscreen propper and since we need a reliable way to turn the pro off I added it staight away. I think a single button hangs it at this point as it expects a second one (and I probably forgot to avoid erroring out), could you either provider `/root/printer_data/logs/guppyscreen.log` or try setting two buttons (an abort) as well like the power button command to test if that is the issue?

**Reply by consp** [reply]
I've written it by trail and error since I could not find the propper description of how the commands work, so it might also be a mismatch between my understanding of the prompt system and what it's supposed to do.

**Reply by consp** [reply]
> If I dismiss via Fluidd, it hangs Guppyscreen

Re-reading your question makes me think it segfaults on a double free since the second button is not initialized.

**Reply by b-fitzpatrick** [reply]
I get it. I also couldn't find anything definitive about the action:prompt messages. Thank you for your efforts. I'm just trying to help with what I can.

I repeated the test with two buttons. Again, it works if I dismiss the dialog from guppyscreen. When I dismiss it from Fluidd, it does not disappear from the screen. When I then tap the OK button on-screen, the button text changes and guppyscreen hangs.
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/3b4fbe83-b660-4959-98e2-218a5c0714e2)

Here's the macro I used:
```
[gcode_macro TEST_PROMPT]
gcode:
    RESPOND TYPE=command MSG="action:prompt_begin Test Prompt"
    RESPOND TYPE=command MSG="action:prompt_text This is a test prompt."
    RESPOND TYPE=command MSG="action:prompt_footer_button Abort|RESPOND TYPE=command MSG=action:prompt_end"
    RESPOND TYPE=command MSG="action:prompt_footer_button OK|RESPOND TYPE=command MSG=action:prompt_end"
    RESPOND TYPE=command MSG="action:prompt_show"
```

Here's the guppyscreen.log:
```
[2024-05-10 04:18:53.316] [guppyscreen] [debug] done init
[2024-05-10 04:19:00.206] [guppyscreen] [debug] data found
[2024-05-10 04:19:00.206] [guppyscreen] [debug] data: // action:prompt_begin Test Prompt
[2024-05-10 04:19:00.206] [guppyscreen] [debug] action: prompt_begin Test Prompt
[2024-05-10 04:19:00.206] [guppyscreen] [debug] PROMPT_BEGIN: Test Prompt
[2024-05-10 04:19:00.694] [guppyscreen] [debug] data found
[2024-05-10 04:19:00.694] [guppyscreen] [debug] data: // action:prompt_text This is a test prompt.
[2024-05-10 04:19:00.694] [guppyscreen] [debug] action: prompt_text This is a test prompt.
[2024-05-10 04:19:00.694] [guppyscreen] [debug] PROMPT_TEXT: This is a test prompt.
[2024-05-10 04:19:01.405] [guppyscreen] [debug] data found
[2024-05-10 04:19:01.405] [guppyscreen] [debug] data: // action:prompt_footer_button Abort|RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:01.405] [guppyscreen] [debug] action: prompt_footer_button Abort|RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:01.405] [guppyscreen] [debug] 26 -1
[2024-05-10 04:19:01.434] [guppyscreen] [debug] PROMPT_FOOTER_BUTTON: Abort CMD: RESPOND TYPE=command MSG=action:prompt_end, type none
[2024-05-10 04:19:02.157] [guppyscreen] [debug] data found
[2024-05-10 04:19:02.157] [guppyscreen] [debug] data: // action:prompt_footer_button OK|RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:02.157] [guppyscreen] [debug] action: prompt_footer_button OK|RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:02.158] [guppyscreen] [debug] 23 -1
[2024-05-10 04:19:02.158] [guppyscreen] [debug] PROMPT_FOOTER_BUTTON: OK CMD: RESPOND TYPE=command MSG=action:prompt_end, type none
[2024-05-10 04:19:02.652] [guppyscreen] [debug] data found
[2024-05-10 04:19:02.652] [guppyscreen] [debug] data: // action:prompt_show
[2024-05-10 04:19:02.653] [guppyscreen] [debug] action: prompt_show
[2024-05-10 04:19:02.653] [guppyscreen] [debug] PROMPT_SHOW
[2024-05-10 04:19:02.653] [guppyscreen] [debug] callback
[2024-05-10 04:19:08.380] [guppyscreen] [debug] Button 1 pressed
[2024-05-10 04:19:08.380] [guppyscreen] [debug] Command issued: RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:08.380] [guppyscreen] [debug] send_jsonrpc: {"id":98,"jsonrpc":"2.0","method":"printer.gcode.script","params":{"script":"RESPOND TYPE=command MSG=action:prompt_end"}}
[2024-05-10 04:19:08.882] [guppyscreen] [debug] data found
[2024-05-10 04:19:08.882] [guppyscreen] [debug] data: // action:prompt_end
[2024-05-10 04:19:08.882] [guppyscreen] [debug] action: prompt_end
[2024-05-10 04:19:08.882] [guppyscreen] [debug] PROMPT_END
[2024-05-10 04:19:16.310] [guppyscreen] [debug] data found
[2024-05-10 04:19:16.310] [guppyscreen] [debug] data: // action:prompt_begin Test Prompt
[2024-05-10 04:19:16.310] [guppyscreen] [debug] action: prompt_begin Test Prompt
[2024-05-10 04:19:16.310] [guppyscreen] [debug] PROMPT_BEGIN: Test Prompt
[2024-05-10 04:19:16.810] [guppyscreen] [debug] data found
[2024-05-10 04:19:16.810] [guppyscreen] [debug] data: // action:prompt_text This is a test prompt.
[2024-05-10 04:19:16.810] [guppyscreen] [debug] action: prompt_text This is a test prompt.
[2024-05-10 04:19:16.810] [guppyscreen] [debug] PROMPT_TEXT: This is a test prompt.
[2024-05-10 04:19:17.524] [guppyscreen] [debug] data found
[2024-05-10 04:19:17.524] [guppyscreen] [debug] data: // action:prompt_footer_button Abort|RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:17.524] [guppyscreen] [debug] action: prompt_footer_button Abort|RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:17.524] [guppyscreen] [debug] 26 -1
[2024-05-10 04:19:17.524] [guppyscreen] [debug] PROMPT_FOOTER_BUTTON: Abort CMD: RESPOND TYPE=command MSG=action:prompt_end, type none
[2024-05-10 04:19:18.240] [guppyscreen] [debug] data found
[2024-05-10 04:19:18.240] [guppyscreen] [debug] data: // action:prompt_footer_button OK|RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:18.240] [guppyscreen] [debug] action: prompt_footer_button OK|RESPOND TYPE=command MSG=action:prompt_end
[2024-05-10 04:19:18.240] [guppyscreen] [debug] 23 -1
[2024-05-10 04:19:18.240] [guppyscreen] [debug] PROMPT_FOOTER_BUTTON: OK CMD: RESPOND TYPE=command MSG=action:prompt_end, type none
[2024-05-10 04:19:18.732] [guppyscreen] [debug] data found
[2024-05-10 04:19:18.732] [guppyscreen] [debug] data: // action:prompt_show
[2024-05-10 04:19:18.732] [guppyscreen] [debug] action: prompt_show
[2024-05-10 04:19:18.733] [guppyscreen] [debug] PROMPT_SHOW
[2024-05-10 04:19:18.733] [guppyscreen] [debug] callback
[2024-05-10 04:19:19.249] [guppyscreen] [debug] data found
[2024-05-10 04:19:19.250] [guppyscreen] [debug] data: // action:prompt_end
[2024-05-10 04:19:19.250] [guppyscreen] [debug] action: prompt_end
[2024-05-10 04:19:19.250] [guppyscreen] [debug] PROMPT_END
[2024-05-10 04:19:28.038] [guppyscreen] [debug] Button 1 pressed
```

**Reply by b-fitzpatrick** [reply]
The same thing happens when I use the SHUTDOWN macro. If I tap ABORT in Fluidd, the prompt remains on screen, and tapping a button will hang it.

For convenience, I've added a macro to restart guppyscreen (in case it's useful to others):
```
[gcode_shell_command restart_guppy]
command: /mnt/data/.klipper_mod/chroot/etc/init.d/S80guppyscreen restart
timeout: 6.
verbose: False

[gcode_macro RESTART_GUPPY]
gcode:
  RUN_SHELL_COMMAND CMD=restart_guppy
```

**Reply by b-fitzpatrick** [reply]
@consp Looking at your code, I think it needs to close the msgbox when receiving a "prompt_end" command. When the "prompt_end" is initiated from Fluidd/Mainsail, it cleans the pdata but leaves the msgbox open.

**Reply by consp** [reply]
> I think it needs to close the msgbox when receiving a "prompt_end" command.

I think you are correct, I'll try to add it tomorrow and see what works and not.

> it cleans the pdata but leaves the msgbox open.

That will definitely throw a segv :roll_eyes: 
I'm more of a c programmer and I'm mixing c/c++ is probably causing some errors. 


Tnx for the spot, I'm making an issue of it for it someone else tries it and doesn't read this thread :)

**Reply by consp** [reply]
>  (but not button groups, yet)

For my reference: https://docs.mainsail.xyz/overview/features/macro-prompts

Button groups will be a challenge as the underlying gui prompt used now in guppy does not really support this (it's a very simple thing). Might need a custom box and other scaffolding not yet in place. Might add that just as well so it can be mainlined in the future. Again: main goal was the shutdown button for now but this is a very useful tool like with the restart/firmware restart prompts on other displays.

**Reply by xblax** [reply]
Just a quick note: we do already use button groups for the LOAD_MATERIAL, M600 commands and in case of filament run out (_FILAMENT_RUNOUT_EVENT). These prompts currently don't work with Guppyscreen. If you plan to implement support for it, you can use these Macros for testing.

**Reply by consp** [reply]
Went with the diy aproach so it will likely be possible (only visuals so far so not there yet):

![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/6802998/7d075340-09ed-49e9-82b9-29ae80e832e2)

At least it won't segv and it's all c++ so portable to master which is always nice. 

These are the prompt footer buttons for now but the idea is the same for button groups.

**Reply by consp** [reply]
For those interested in the attachment the a guppyscreen file with the [latest changes](https://github.com/ballaswag/guppyscreen/compare/main...consp:guppyscreen:flashforge_ad5m), no warenty whatsoever, replace `/root/printer_software/guppyscreen/guppyscreen` with it and restart guppyscreen or reboot (rename, because github).

> [!CAUTION]
> If you do not know how to do the above this is not meant for you. Bugs may happen.

I'm trying it out now, let's see how it goes.

[guppyscreen.txt](https://github.com/xblax/flashforge_ad5m_klipper_mod/files/15310130/guppyscreen.txt)

![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/6802998/891f2b17-ed82-4c05-b8b6-3e8c6ee5ec62)

**Reply by b-fitzpatrick** [reply]
@consp has it been working well for you? Any fixes coming? If this is still the latest and greatest, I'll give it a try.

**Reply by KaruroChori** [reply]
I used it to handle some macros to switch material mid print and it works well for me.

**Reply by consp** [reply]
So far it works fine for me, but I mainly use it now for shutting down now (for the "shutdown" prompt).

No errors yet and I've abused it quite a bit. Haven't checked for memory leaks but it's also mainlined now so if those occur it will definitely show up somewhere.

Sidenote: The spelling mistake in my test prompt is starting to annoy me ... :man_facepalming:

**Reply by b-fitzpatrick** [reply]
So far, so good

**Reply by johncotten** [reply]
@consp and @ballaswag  Thank you for fixing this guppy screen problem. I can confirm that version of guppy screen you attaches works on v.05 and firmware 2.7.8. Could you commit to v.05 or move to v.06? 

Looks like everyone's hard work is getting noticed:
https://www.youtube.com/watch?v=AWafigSeXf8

**Comment by b-fitzpatrick** [comment]
This isn’t new, but in case you haven’t noticed, it appears there’s some sort of status message that’s mostly off-screen. It shows up on the home screen when a print job is paused.
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/9221473/ee15a7df-910a-4127-9680-812654bd934a)

**Reply by consp** [reply]
Ah yes, probably due to the change in location of the temperature numers, which needed to move a bit if you want the pressure and air sensors in there as well.

Should be fixable, I'll have a look.

**Reply by ballaswag** [reply]
> Ah yes, probably due to the change in location of the temperature numers, which needed to move a bit if you want the pressure and air sensors in there as well.
> 
> Should be fixable, I'll have a look.

Maybe just make that container scrollable , then more sensors can be displayed.

**Reply by consp** [reply]
> Maybe just make that container scrollable , then more sensors can be displayed.

That is an option but it looks horrible as the last 20% of the bottom one is what is cut off and there is no obvious way to display it's scrollable without a grey bar in the middle of the screen.
